### PR TITLE
RHOAIENG-31273: Add Cypress mock tests for testing gaps

### PR DESCRIPTION
**Pull request by rsun19**
https://redhat.atlassian.net/browse/RHOAIENG-31273

## Description

Adds new Cypress mock tests to close testing gaps identified across multiple linked Jira tickets:

| Ticket | Area | What's tested |
|--------|------|---------------|
| [RHOAIENG-18214](https://redhat.atlassian.net/browse/RHOAIENG-18214) / [RHOAIENG-17122](https://redhat.atlassian.net/browse/RHOAIENG-17122) | Workbench env vars | Pre-existing `envFrom` ConfigMap/Secret references persist correctly when editing a notebook; removal of individual entries |
| [RHOAIENG-15228](https://redhat.atlassian.net/browse/RHOAIENG-15228) | Data connections | Workbench loads without false "env var not found" warnings when connections are present |
| [RHOAIENG-5349](https://redhat.atlassian.net/browse/RHOAIENG-5349) | Pipeline graph | Complex topologies render correctly (fan-out/fan-in, diamond, sequential chains, large 20-task graphs) |
| [RHOAIENG-8635](https://redhat.atlassian.net/browse/RHOAIENG-8635) | Pipeline server | Delete button visibility based on user role, delete confirmation modal, deletion flow, and server state handling |
| [RHOAIENG-6317](https://redhat.atlassian.net/browse/RHOAIENG-6317) | Pipeline logs | Error handling for 404, 500, empty, and malformed log responses; multi-container log retrieval |
| [RHOAIENG-7430](https://redhat.atlassian.net/browse/RHOAIENG-7430) | Pipeline toolbar | Collapse/expand all actions in the pipeline topology toolbar |
| [RHOAIENG-6851](https://redhat.atlassian.net/browse/RHOAIENG-6851) | AI Models tab | Kebab menu "Remove asset" action — modal flow, loading state, error with retry, success with list refresh, multi-model scenarios, accessibility |

**Supporting changes:**
- New page object methods in `topology.ts` (toolbar buttons, node/edge finders), `workbench.ts` (scrollIntoView for connection rows), `pipelinesGlobal.ts` (public server action button), and `modelsTabPage.ts` (`DeleteModelModal`, `KebabMenu` classes)
- `mock500Error` helper added to `mockK8sStatus.ts`
- `data-testid="model-actions-kebab"` added to `AIModelTableRow.tsx` for reliable test selection

## How Has This Been Tested?

All new and modified Cypress mock tests were run locally in headed mode (`npx cypress open`) and verified to pass. Existing tests in the affected files were confirmed to still pass after the changes.

## Test Impact

This PR is entirely new test coverage — 13 files changed, ~1,930 lines of new mock tests and page object support. No application logic changes beyond adding a single `data-testid` attribute.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [x] The code follows our [Best Practices](/docs/best-practices.md) (React coding standards, PatternFly usage, performance considerations)

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: packages/cypress/cypress/tests/mocked/projects/workbenchDataConnectionWarnings.cy.ts [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    // Mock individual connection GET requests
+    connectionNames.forEach((name) => {
+      cy.interceptK8s(
+        { model: SecretModel, ns: 'test-project', name },
+        mockSecretK8sResource({
+          name,
+          namespace: 'test-project',
+          displayName: name,
+          connectionType: 's3',
+        }),
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/projects/workbenchDataConnectionWarnings.cy.ts:139]: Intentionally kept as-is. This test file validates connection **warning behavior** (missing connections, missing env var secrets, annotation-based vs envFrom-based), not connection-type-specific behavior. All connections in these tests are intentionally s3 because the warning logic is type-agnostic — a missing connection triggers a warning regardless of whether it is s3, postgres, or mysql. Changing the mock type per connection would add complexity without improving coverage for this file's purpose.

File: packages/cypress/cypress/pages/workbench.ts [hunk 2]
```diff
@@ -684,8 +684,9 @@ class CreateSpawnerPage {
   }
 
   findConnectionsTableRow(name: string, type: string) {
-    this.findConnectionsTable().find(`[data-label=Name]`).contains(name);
-    this.findConnectionsTable().find(`[data-label=Type]`).contains(type);
+    this.findConnectionsTable().scrollIntoView();
+    this.findConnectionsTable().find(`[data-label=Name]`).contains(name).should('be.visible');
+    this.findConnectionsTable().find(`[data-label=Type]`).contains(type).should('be.visible');
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/pages/workbench.ts]: Fixed — `findConnectionsTableRow` now locates the `<tr>` via `.parents('tr')` and uses `.within()` to scope both Name and Type assertions to the same row.

File: packages/cypress/cypress/tests/mocked/pipelines/pipelineGraphEdgeCases.cy.ts [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+      pipelineDetails.findTaskNode('task-a').should('exist');
+      pipelineDetails.findTaskNode('task-b1').should('exist');
+      pipelineDetails.findTaskNode('task-b2').should('exist');
+      pipelineDetails.findTaskNode('task-b3').should('exist');
+      pipelineDetails.findTaskNode('task-c').should('exist');
+
+      pipelineDetails.findTaskNode('task-c').click();
+      const taskDrawer = pipelineDetails.getTaskDrawer();
+      taskDrawer.shouldHaveTaskName('task-c');
+    });
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/pipelines/pipelineGraphEdgeCases.cy.ts:171]: Fixed — added `findEdges().should('have.length', 6)` to fan-out/fan-in and `findEdges().should('have.length', 3)` to sequential dependencies.

File: packages/cypress/cypress/tests/mocked/pipelines/pipelineGraphEdgeCases.cy.ts [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+      );
+
+      pipelineDetails.visit(
+        projectId,
+        mockPipeline.pipeline_id,
+        mockEmptyPipeline.pipeline_version_id,
+      );
+
+      pipelineDetails.findPageTitle().should('exist');
+    });
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/pipelines/pipelineGraphEdgeCases.cy.ts]: Fixed — incomplete spec now asserts `findVisualizationSurface().should('be.visible')` + `findTaskNodes().should('have.length', 1)`. Empty spec asserts the same surface + `findTaskNodes().should('have.length', 0)`.

File: packages/cypress/cypress/tests/mocked/pipelines/pipelines.cy.ts [hunk 5]
```diff
@@ -520,6 +521,159 @@ describe('Pipelines', () => {
...
+
+    it('should disable pipeline server actions when server is not configured', () => {
+      asProductAdminUser();
+      initIntercepts({ isEmpty: true });
+      pipelinesGlobal.visit(projectName);
+
+      pipelinesGlobal.findEmptyState().should('exist');
+      pipelinesGlobal.findPipelineServerActionButton().should('exist').should('be.disabled');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/pipelines/pipelines.cy.ts]: Fixed — added `afterEach(() => { Cypress.env('USER_CONFIG', undefined); })` to the describe block.

File: packages/cypress/cypress/tests/mocked/pipelines/pipelines.cy.ts [hunk 6]
```diff
@@ -520,6 +521,159 @@ describe('Pipelines', () => {
...
+      pipelinesGlobal.selectPipelineServerAction('Delete pipeline server');
+      deleteModal.shouldBeOpen();
+    });
+
+    it('should show delete option when server has error', () => {
+      initIntercepts({ initializing: true, errorMessage: 'Failed to connect to storage' });
+      pipelinesGlobal.visit(projectName);
+
+      pipelinesGlobal.selectPipelineServerAction('Delete pipeline server');
+      deleteModal.shouldBeOpen();
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/pipelines/pipelines.cy.ts]: Fixed — changed `initializing: true` to `initializing: false` so the test exercises the pure error branch.

File: packages/cypress/cypress/tests/mocked/pipelines/topology/pipelinesTopology.cy.ts [hunk 7]
```diff
@@ -821,5 +822,190 @@ describe('Pipeline topology', () => {
...
+          },
+        },
+        '\x00\x01\x02malformed\nlog\ndata',
+      ).as('malformedLogs');
+
+      navigateToLogsTab();
+
+      cy.wait('@malformedLogs');
+      pipelineRunDetails.findLogs().should('be.visible');
+    });
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/pipelines/topology/pipelinesTopology.cy.ts]: Fixed — all 4 edge-case tests (404, 500, empty, malformed) now assert `findLogsSuccessAlert().should('not.exist')`. Empty logs also asserts `findLogs().contains('No logs available')`.

File: packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+      editSpawnerPage
+        .getEnvironmentVariableTypeField(0)
+        .findRemoveEnvironmentVariableButton()
+        .click();
+
+      editSpawnerPage
+        .getEnvironmentVariableTypeField(0)
+        .find()
+        .scrollIntoView()
+        .should('be.visible');
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts:152]: Fixed — added `findEnvironmentVariableFields().should('have.length', N)` after each removal in both test blocks, plus a new `findEnvironmentVariableFields()` page object method.

File: packages/gen-ai/frontend/src/__tests__/cypress/cypress/tests/mocked/aiAssets/modelTabKebabActions.cy.ts [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+      kebabMenu.findRemoveAssetItem().click();
+      deleteModelModal.findRemoveButton().click();
+
+      cy.step('Verify loading state');
+      deleteModelModal.findRemovingButton().should('be.visible').and('be.disabled');
+      deleteModelModal.findCancelButton().should('be.disabled');
+
+      cy.step('Wait for deletion to complete');
+      cy.wait('@deleteModel');
+    },
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/gen-ai/frontend/src/__tests__/cypress/cypress/tests/mocked/aiAssets/modelTabKebabActions.cy.ts]: Fixed — renamed to "should display loading state while deleting". The error+retry flow is fully covered in the separate test at line 173.

File: packages/gen-ai/frontend/src/__tests__/cypress/cypress/tests/mocked/aiAssets/modelTabKebabActions.cy.ts [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+      modelsTabPage
+        .getRow('Custom GPT Endpoint')
+        .findKebabMenu()
+        .should('have.attr', 'aria-label', 'Actions for Custom GPT Endpoint');
+
+      cy.step('Open menu and verify menu roles');
+      modelsTabPage.getRow('Custom GPT Endpoint').findKebabMenu().click();
+      kebabMenu.shouldBeVisible();
+      kebabMenu.findRemoveAssetItem().should('exist');
+    },
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/gen-ai/frontend/src/__tests__/cypress/cypress/tests/mocked/aiAssets/modelTabKebabActions.cy.ts:362]: Fixed — added Enter and Space key activation tests. The test now opens the menu via mouse click, then via `.focus().type('{enter}')`, then via `.focus().type(' ')`, each asserting the menu opens and Remove asset item exists.

File: packages/gen-ai/frontend/src/__tests__/cypress/cypress/tests/mocked/aiAssets/modelTabKebabActions.cy.ts [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+      cy.interceptGenAi('GET /api/v1/aaa/models', { data: [CUSTOM_MODEL_2, NAMESPACE_MODEL] }).as(
+        'refreshModels',
+      );
+
+      deleteModelModal.findRemoveButton().click();
+
+      cy.wait('@deleteModel1');
+      cy.wait('@refreshModels');
+
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/gen-ai/frontend/src/__tests__/cypress/cypress/tests/mocked/aiAssets/modelTabKebabActions.cy.ts:268]: Fixed — added `cy.wait('@deleteModel1').then((interception) => { expect(interception.request.url).to.include('model_id=custom-model-1'); })` to assert the correct model identifier is sent in the delete request.

File: packages/cypress/cypress/tests/mocked/pipelines/pipelineGraphEdgeCases.cy.ts [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+
+    it('should render pipeline with 20+ tasks', () => {
+      pipelineDetails.visit(
+        projectId,
+        mockPipeline.pipeline_id,
+        mockLargePipeline.pipeline_version_id,
+      );
+
+      pipelineDetails.findTaskNodes().should('have.length', 25);
+    });
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/pipelines/pipelineGraphEdgeCases.cy.ts:248]: Fixed — added `findEdges().should('have.length', 24)` to assert all 24 sequential dependency edges render alongside the 25 task nodes.

File: packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        .getEnvironmentVariableTypeField(1)
+        .find()
+        .scrollIntoView()
+        .should('be.visible');
+      editSpawnerPage
+        .getEnvironmentVariableTypeField(2)
+        .find()
+        .scrollIntoView()
+        .should('be.visible');
+
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts]: Fixed — added `findEnvironmentVariableFields().should('have.length', 3)` before the per-index visibility checks.

File: packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+      envField.selectEnvDataType('Key / value');
+      envField.getKeyValuePair(0).findKeyInput().fill('SECOND_KEY');
+      envField.getKeyValuePair(0).findValueInput().fill('second_value');
+
+      editSpawnerPage
+        .getEnvironmentVariableTypeField(0)
+        .getKeyValuePair(0)
+        .findKeyInput()
+        .should('have.value', 'SECOND_KEY');
+    });
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts:266]: Fixed — added `findValueInput().should('have.value', 'second_value')` assertion alongside the existing key input check.

File: packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+        .findRemoveEnvironmentVariableButton()
+        .click();
+
+      editSpawnerPage.findEnvironmentVariableFields().should('have.length', 1);
+      editSpawnerPage
+        .getEnvironmentVariableTypeField(0)
+        .find()
+        .scrollIntoView()
+        .should('be.visible');
+    });
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/cypress/cypress/tests/mocked/projects/environmentVariableEdgeCases.cy.ts:153]: Fixed — added `findEnvironmentVariableFields().should('have.length', 0)` after removing the last variable and before re-adding. Also added initial count assertion of 3 for the pre-existing env vars block.

File: packages/gen-ai/frontend/src/__tests__/cypress/cypress/pages/modelsTabPage.ts [hunk 16]
```diff
@@ -62,6 +66,58 @@ class ModelsTabPage {
...
+  find(): Cypress.Chainable<JQuery<HTMLElement>> {
+    return cy.findByTestId('delete-model-modal');
+  }
+
+  shouldBeVisible(): void {
+    this.find().should('be.visible');
+  }
+
+  shouldNotExist(): void {
+    this.find().should('not.exist');
```
[the remark this answers is not in the document]
Comment (rsun19) [reply packages/gen-ai/frontend/src/__tests__/cypress/cypress/pages/modelsTabPage.ts:79]: This was investigated and attempted earlier in this PR, then reverted. `@testing-library/cypress` v10 (the version installed in this project) does not expose `queryBy*` commands — only `findBy*` and `findAllBy*` are available on `cy`. The existing `this.find().should('not.exist')` pattern works correctly: when `findBy*` is chained with `.should('not.exist')`, Cypress retries until the element is absent (or times out), rather than failing on first find. This is a well-established convention used throughout the project.

File: packages/cypress/cypress/tests/mocked/pipelines/pipelines.cy.ts [hunk 17]
```diff
@@ -520,6 +521,163 @@ describe('Pipelines', () => {
...
+      pipelinesGlobal.findPipelineServerActionButton().should('exist').should('be.enabled');
+    });
+
+    it('should show pipeline server actions for project admin users', () => {
+      asProjectAdminUser();
+      initIntercepts({});
+      pipelinesGlobal.visit(projectName);
+
+      pipelinesGlobal.findPipelineServerActionButton().should('exist');
+    });
```
Comment (christianvogt) [inline packages/cypress/cypress/tests/mocked/pipelines/pipelines.cy.ts]: This test has the exact same name but do different assertions.

We need to aim for efficiency and bundle assertions together where the setup is identical because every test requires a page load from scratch and those times add up in cypress tests.

Comment (christianvogt) [reply packages/cypress/cypress/tests/mocked/pipelines/pipelines.cy.ts]: haha ok so one is project and another is product :( i guess my brain didn't notice the difference when scanning.

File: packages/cypress/cypress/pages/pipelines/topology.ts [hunk 18]
```diff
@@ -63,6 +63,47 @@ class PipelinesTopology {
...
+    return cy.get('[data-kind="node"]');
+  }
+
+  findEdges() {
+    return cy.get('[data-kind="edge"]');
+  }
+
+  // PatternFly Topology third-party component — no data-testid support
+  findVisualizationSurface() {
+    return cy.get('.pf-topology-visualization-surface');
```
Comment (bobbravo2) [inline packages/cypress/cypress/pages/pipelines/topology.ts]: do we own this selector/xpath? If yes, LGTM, if not, we should wrap it so this doesn't break on us in the future when PF upgrades.

File: packages/cypress/cypress/pages/workbench.ts [hunk 19]
```diff
@@ -684,8 +688,15 @@ class CreateSpawnerPage {
   }
 
   findConnectionsTableRow(name: string, type: string) {
-    this.findConnectionsTable().find(`[data-label=Name]`).contains(name);
-    this.findConnectionsTable().find(`[data-label=Type]`).contains(type);
+    this.findConnectionsTable().scrollIntoView();
```
Comment (bobbravo2) [inline packages/cypress/cypress/pages/workbench.ts:691]: scrolling _**might be**_ fragile

Comment (bobbravo2) [reply packages/cypress/cypress/pages/workbench.ts:691]: Any way to wait until an element is present or scroll an element into view?

File: packages/cypress/cypress/tests/mocked/pipelines/topology/pipelinesTopology.cy.ts [hunk 20]
```diff
@@ -821,5 +822,166 @@ describe('Pipeline topology', () => {
...
+        },
+        mock500Error({}),
+      ).as('logs500');
+
+      navigateToLogsTab();
+      cy.wait('@logs500');
+      pipelineRunDetails.findLogs().should('be.visible');
+      pipelineRunDetails.findLogsSuccessAlert().should('not.exist');
+    });
+
```
Comment (bobbravo2) [inline packages/cypress/cypress/tests/mocked/pipelines/topology/pipelinesTopology.cy.ts:873]: Thank you. 500 / ISEs are a huge gap that this work fills for us!

**Review by bobbravo2 — CHANGES_REQUESTED** [review]
TLDR: Great work iteratively improving coverage. There are a few cleanup / idempotency gaps that will help us ensure that this code cleans up after itself. The `a11y` tests could also use some keyboard coverage, but that's a product gap and can wait for a followup PR. 

Reviewed with Opus 4.6. Prompt: "Flakes are blockers" 

**The three flake vectors that should block merge:**

1. Role test state leak. The merged it() block chains 3 role transitions with inline Cypress.env('USER_CONFIG', undefined) resets. If any assertion throws mid-block, cleanup never runs and USER_CONFIG leaks into every downstream test in the suite. This needs either separate it() blocks per role or an **_afterEach_** guard on the enclosing describe.

2. Log error tests are false-positive-prone. Asserting findLogs().should('be.visible') on 404/500/malformed responses proves the panel rendered, not that the **error path executed**. If someone breaks the error handler and the **panel renders empty** or with stale content, **these tests pass green**. The empty-log test already does contains('No logs available') -- the 404/500/malformed tests need equivalent positive assertions on the error UI state (error text, error alert, or specific empty-state indicator).

3. Global cy.findByRole('menu') in KebabMenu class. The multi-model test opens kebab menus for different rows in sequence. If a close animation hasn't completed before the next kebabMenu.shouldBeVisible(), the query binds to a stale DOM node. This needs to be scoped to the triggering row's data-testid, or at minimum chained after an explicit should('not.exist') wait on the previous menu.

**All three are deterministic-in-CI-until-they-aren't patterns. They won't fail in local headed mode but will eventually flake in parallel CI runs or under load.**

File: packages/cypress/cypress/pages/workbench.ts (hunk 19 above)
Comment (rsun19) [reply packages/cypress/cypress/pages/workbench.ts:691]: Is this not waiting until an element is scrolled into view?

File: packages/cypress/cypress/pages/pipelines/topology.ts (hunk 18 above)
Comment (rsun19) [reply packages/cypress/cypress/pages/pipelines/topology.ts]: Right now, PF is using `<div data-test-id="topology" className={css(styles.topologyVisualizationSurface)} ref={measureRef}>`. If not fixed yet we can open an issue so it is `data-testid` instead so that we can use our implementation of `cy.findByTestId()` method

File: packages/cypress/cypress/pages/workbench.ts (hunk 19 above)
Comment (bobbravo2) [reply packages/cypress/cypress/pages/workbench.ts:691]: you're right. My mistake!

**Review by bobbravo2 — APPROVED** [review]
Thank you!
